### PR TITLE
Add user interface for iPad

### PR DESCRIPTION
**Pull request by pronebird**
Describe **what** this PR changes. **Why** this is wanted. And, if needed, **how** it does it.

Git checklist:

* [X] Describe the change in **`CHANGELOG.md`** under the `[Unreleased]` header.
* [X] Check that commits follow the [Mullvad coding guidelines](https://github.com/mullvad/coding-guidelines)

This PR adds split view controller UI on iPad and makes sure that we properly handle split screen on iPad where the app would resort to iPhone UI when in compact mode.

---
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)

### CODE REVIEW

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, all discussions resolved

___
*[ios/MullvadVPN/AppDelegate.swift, line 238 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwKG8MBqCUWoHZyh22:-MZwKG8NA4dqUXlTUzFI:brcdlzu) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L238)):*
> ```Swift
>             consentViewController.modalPresentationStyle = .formSheet
>             if #available(iOS 13.0, *) {
>                 consentViewController.isModalInPresentation = true
> ```

FYI makes sure that the controller cannot be dismissed with swipe.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, all discussions resolved

___
*[ios/MullvadVPN/AppDelegate.swift, line 310 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwK_PR2Kw8-yXvUzPq:-MZwK_PR2Kw8-yXvUzPr:bz7x5ka) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L310)):*
> ```Swift
>     }
> 
>     private func showSplitViewMaster(_ flag: Bool, animated: Bool) {
> ```

FYI Agreed with Matilda that we'll be hiding the split view with all content except the world map before user logged in. So this helper does exactly that.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, all discussions resolved

___
*[ios/MullvadVPN/AppDelegate.swift, line 548 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwL6sA8sFiw4Zwabaa:-MZwL6sA8sFiw4Zwabab:bod12sv) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L548)):*
> ```Swift
>     func selectLocationViewController(_ controller: SelectLocationViewController, didSelectRelayLocation relayLocation: RelayLocation) {
>         // Dismiss view controller in modal presentation
>         if controller.presentingViewController != nil {
> ```

FYI this delegate handles different presentations of select location controller:

1. Modal presentation, used on iPhone or iPad in compact horizontal environment.
2. Inline controller presented as a part of split controller on iPad

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, all discussions resolved

___
*[ios/MullvadVPN/AppDelegate.swift, line 587 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwLn2zDhRp_I5-r-7W:-MZwLn2zDhRp_I5-r-7X:bf0l45q) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L587)):*
> ```Swift
> extension AppDelegate: UIAdaptivePresentationControllerDelegate {
> 
>     func adaptivePresentationStyle(for controller: UIPresentationController, traitCollection: UITraitCollection) -> UIModalPresentationStyle {
> ```

FYI this delegate method handles the presentation in different environments on iPad. We tend to keep controllers smaller in regular environment as there is a lot of screen estate and then make them fullscreen in compact environments which makes the iPad UI look exactly the same way it looks on iPhone without any other changes.
___
*[ios/MullvadVPN/AppDelegate.swift, line 602 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwM68h0Ne-DPzMG49S:-MZwM68h0Ne-DPzMG49T:bmwhxpx) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L602)):*
> ```Swift
> 
>     func presentationController(_ controller: UIPresentationController, viewControllerForAdaptivePresentationStyle style: UIModalPresentationStyle) -> UIViewController? {
>         return nil
> ```

FYI this allows wrapping or replacing the view controller based on presentation style, however this didn't quite work for modal controllers and I had different issues with keyboard, so instead we wrap some of controllers before hand and then just hide the undesired UI elements. See the `makeLoginContainerController()`.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/AppDelegate.swift, line 608 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwMYGQ-EBd4IKJV6h4:-MZwMYGQ-EBd4IKJV6h5:b-tuia0n) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L608)):*
> ```Swift
>         // Force hide header bar in .formSheet presentation and show it in .fullScreen presentation
>         if let wrapper = presentationController.presentedViewController as? RootContainerViewController {
>             wrapper.setOverrideHeaderBarHidden(style == .formSheet, animated: false)
> ```

And this is related to the comment above. We need to hide certain UI elements in regular environments but then show them in compact environments and because modifying the view hierarchy is kinda buggy or didn't quite work, instead we override couple of things in here.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/ConnectViewController.swift, line 269 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwQ1Bc6Mfp-8I4ntKL:-MZwQ1Bc6Mfp-8I4ntKM:bczjfp2) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/ConnectViewController.swift#L269)):*
> ```Swift
>             }
> 
>         case (.pad, .regular):
> ```

FYI I don't show the "select location" on iPad when in split view since the relay picker is displayed side-by-side.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/AppDelegate.swift, line 608 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwMYGQ-EBd4IKJV6h4:-MZwTVQGDufheR8PYnpD:b8isiah) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L608)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

And this is related to the comment above. We need to hide certain UI elements in regular environments but then show them in compact environments and because modifying the view hierarchy is kinda buggy or didn't quite work, instead we override couple of things in here.
</blockquote></details>

Basically on iPad we always have the login controller wrapped in our custom container that has the Mullvad logo and settings button. However we hide the header bar on iPad in regular environment. On the screenshot below the header bar is visible but that's because I am in UI inspection mode, otherwise it's invisible.




and then show it when in compact environment and with help of adaptive presentation we make this controller fullscreen which makes it look exactly the same way it looks on iPhone:

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/ConnectViewController.swift, line 84 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwZLKtEB3Sh53KctzR:-MZwZLKtEB3Sh53KctzS:brfcis6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/ConnectViewController.swift#L84)):*
> ```Swift
>         if previousTraitCollection?.userInterfaceIdiom != traitCollection.userInterfaceIdiom ||
>             previousTraitCollection?.horizontalSizeClass != traitCollection.horizontalSizeClass {
>             updateTraitDependentViews()
> ```

FYI this code makes sure that we adapt the UI to the changes in environment. For instance in split screen environment our app may become very slim which should collapse the UI to look like on iPhone.

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/ConnectViewController.swift, line 269 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwQ1Bc6Mfp-8I4ntKL:-MZwZoCc3MZfIllekdnG:b-xe4tda) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/ConnectViewController.swift#L269)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

FYI I don't show the "select location" on iPad when in split view since the relay picker is displayed side-by-side.
</blockquote></details>

%20-%202021-05-05%20at%2014.49.10.png)
%20-%202021-05-05%20at%2014.49.37.png)

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 0 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/CustomSplitViewController.swift, line 63 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-MZwkHiB0HuwaZvEFXHv:-MZwkHiC-XRbhadC91R0:be944d6) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/CustomSplitViewController.swift#L63)):*
> ```Swift
>         // view is expanded.
>         if !self.isCollapsed, childViewController == self.viewControllers.last {
>             let sizeOverrideTraitCollection = UITraitCollection(horizontalSizeClass: self.traitCollection.horizontalSizeClass)
> ```

FYI This basically overrides the horizontal class for the environment of child controllers when in split view.

**Review by pinkisemils — COMMENTED** [review]
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 1 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/AppDelegate.swift, line 310 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_09MXRCj_HhW-4Kjab:-M_09MXRCj_HhW-4Kjac:bqqyk9i) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L310)):*
> ```Swift
>     }
> 
>     private func showSplitViewMaster(_ flag: Bool, animated: Bool) {
> ```

`flag` is a really non-descript name for this variable, IMO.
___
*[ios/MullvadVPN/AppDelegate.swift, line 325 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_09cGq5W3G9bsIlX9G:-M_09cGq5W3G9bsIlX9H:b5y3fa4) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L325)):*
> ```Swift
> 
> extension AppDelegate: RootContainerViewControllerDelegate {
> ```

Is there a not a swift formater that'd eliminate whitespace changes?

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 1 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/AppDelegate.swift, line 310 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_09MXRCj_HhW-4Kjab:-M_0FjbWCvFyRNkOoNxO:b-1yibzn) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L310)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

`flag` is a really non-descript name for this variable, IMO.
</blockquote></details>

The name of the function implies the meaning of the first argument, hence I use `_` to tell the compiler to omit the first argument name in call signature, i.e:

```
showSplitViewMaster(true, animated: false)
```

`flag` is basically an internal alias. I thought it was good enough.
___
*[ios/MullvadVPN/AppDelegate.swift, line 325 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_09cGq5W3G9bsIlX9G:-M_0FXSy4Dz1Hh01djG9:b-k3vx1z) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L325)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Is there a not a swift formater that'd eliminate whitespace changes?
</blockquote></details>

I think that's me who added the newline. 

Xcode has built in formatter but it does not eliminate empty newlines or extra whitespace and formatting the code proved to be counter productive because Xcode tends to format it the way it wants, even if often enough the formatting looks odd.

**Review by pinkisemils — COMMENTED** [review]
> Reviewed 1 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 2 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/AppDelegate.swift, line 161 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_0REf56mOqvq7xEgML:-M_0REf56mOqvq7xEgMM:b-sffnme) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L161)):*
> ```Swift
>         if !Account.shared.isAgreedToTermsOfService {
>             let consentViewController = self.makeConsentController { [weak self] (viewController) in
>                 guard let self = self else { return }
> ```

Does this upgrade the weak reference into a strong one? If not, is the weak reference guaranteed to have a valid underlying reference for the lifetime of this function?
___
*[ios/MullvadVPN/ConnectViewController.swift, line 117 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_0Pw-F2TAaX02Np5MN:-M_0Pw-GBN8I7E7Lu1f-:b-j4k800) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/ConnectViewController.swift#L117)):*
> ```Swift
> 
>     func account(_ account: Account, didUpdateExpiry expiry: Date) {
>         // no-op
> ```

Why is this useful?

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 2 of 6 files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/AppDelegate.swift, line 161 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_0REf56mOqvq7xEgML:-M_5hBjpEH5MG4XlOfz3:b-wpmh1i) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L161)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Does this upgrade the weak reference into a strong one? If not, is the weak reference guaranteed to have a valid underlying reference for the lifetime of this function?
</blockquote></details>

Yes, the assignment basically creates a strong reference.

The underlying reference behind the weak reference does not really become invalid, but rather when dereferencing the weak reference, it may return `nil` and there are no guarantees when that may happen, that's why we create the strong reference first thing in the scope of a given closure.
___
*[ios/MullvadVPN/ConnectViewController.swift, line 117 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_0Pw-F2TAaX02Np5MN:-M_5iaNP3RxeHg4UjJ-F:b-xu94g2) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/ConnectViewController.swift#L117)):*
<details><summary><i>Previously, pinkisemils (Emīls Piņķis) wrote…</i></summary><blockquote>

Why is this useful?
</blockquote></details>

Part of `AccountObserver` protocol adoption. Even if we don't really do anything inside this method, we are still obliged to have it to satisfy the protocol requirement.

**Review by jvff — APPROVED** [review]
> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: all files reviewed, 1 unresolved discussion

___
*[ios/MullvadVPN/AppDelegate.swift, line 310 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_09MXRCj_HhW-4Kjab:-M_60VB2CBK4TCdA9i6K:b1l6z0m) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L310)):*
<details><summary><i>Previously, pronebird (Andrej Mihajlov) wrote…</i></summary><blockquote>

The name of the function implies the meaning of the first argument, hence I use `_` to tell the compiler to omit the first argument name in call signature, i.e:

```
showSplitViewMaster(true, animated: false)
```

`flag` is basically an internal alias. I thought it was good enough.
</blockquote></details>

Hmm, maybe `_ show` is a bit clearer in this case? So that the `if` becomes `if show`?

Another option would be to split in two methods: `showSplitViewMaster` and `hideSplitViewMaster`?

**Review by jvff — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: all files reviewed, 1 unresolved discussion (waiting on @pronebird)

**Review by pinkisemils — APPROVED** [review]
> Reviewed 2 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: all files reviewed, 1 unresolved discussion (waiting on @pronebird)

**Review by jvff — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-:-M_QJzDU6wyMPIBNACrl:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by pronebird — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723)* status: 4 of 6 files reviewed, all discussions resolved (waiting on @jvff and @pinkisemils)

___
*[ios/MullvadVPN/AppDelegate.swift, line 310 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/2723#-M_09MXRCj_HhW-4Kjab:-M_QWsEN-o3s8urWLwbH:b-oy9ebu) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d9c6451024a23bb2ebd7fa6e68ef80a54d78e55b/ios/MullvadVPN/AppDelegate.swift#L310)):*
<details><summary><i>Previously, jvff (Janito Vaqueiro Ferreira Filho) wrote…</i></summary><blockquote>

Hmm, maybe `_ show` is a bit clearer in this case? So that the `if` becomes `if show`?

Another option would be to split in two methods: `showSplitViewMaster` and `hideSplitViewMaster`?
</blockquote></details>

I think `flag` is fine. Boolean flag in the function that implies what it does. Either way I renamed `flag` to `show` to comply with the request.
